# Metrics
   
Deux metrics seront utilisés pour testé le modèle: Intersection-Over-Union (IoU, Jaccard Index) and the F1 score (Dice coefficient) 
Références:
* <https://www.jeremyjordan.me/evaluating-image-segmentation-models/>
* <https://towardsdatascience.com/metrics-to-evaluate-your-semantic-segmentation-model-6bcb99639aa2>
* <https://ilmonteux.github.io/2019/05/10/segmentation-metrics.html>
* <https://gist.github.com/ilmonteux/8340df952722f3a1030a7d937e701b5a>
* <https://stats.stackexchange.com/questions/273537/f1-dice-score-vs-iou>
 
## Intersection-Over-Union (IoU, Jaccard Index)
Références: 
* <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_score.html>

## F1 score (Dice coefficient)
Références:
* <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html>

## Premier script de test 
### Python 3.6 virtual environment 


#### Python env 3.6
* Python 3.6
* Création de l'environnement Python virtuel
* Installation des modules Python requis
```
module load python/3.6

# clean the python virtual pyenv36
rm -rf $SLURM_TMPDIR/pyenv36

# create the virtual env
virtualenv --no-download $SLURM_TMPDIR/pyenv36

# activate the python virtual env
source $SLURM_TMPDIR/pyenv36/bin/activate

# check the python and pip version (3.6)
python --version
pip --version

# install requirements
pip3 install --no-index scikit-learn
pip3 install --no-index scikit-image
pip3 install --no-index keras
pip3 install --no-index tensorflow_gpu
```

Préparation des images
```
from skimage import io
import numpy as np
import os

filename = os.path.join('/home/vincelf/DeepLabv3FineTuning-master/CrackForest/Images', '001.jpg')
im = io.imread(filename)
im.shape
# (320, 480, 3)
im2 = np.expand_dims(im,0)
im2.shape
# (1, 320, 480, 3)

filename = os.path.join('/home/vincelf/DeepLabv3FineTuning-master/CrackForest/Masks', '001_label.PNG')
mask = io.imread(filename)
mask.shape
# (320, 480)
m = np.array(mask)
m2 = np.stack((m,)*3,-1)
m2.shape
# (320, 480, 3)
m3 = np.expand_dims(m2,0)
m3.shape
# (1, 320, 480, 3)
```

### Test 1st implementation
Reference: <https://www.jeremyjordan.me/evaluating-image-segmentation-models/>
```
def iou_score(y_true, y_pred):
  intersection = np.logical_and(y_true, y_pred)
  union = np.logical_or(y_true, y_pred)
  iou_score = np.sum(intersection) / np.sum(union)
  return iou_score

iou_score(im2,m3)
# 0.011920572916666667
```

### Test 2nd implementation
Reference: <https://towardsdatascience.com/metrics-to-evaluate-your-semantic-segmentation-model-6bcb99639aa2>

Retourne une erreur
```
# Both y_true and y_pred are m x r x c x n where m is the number of images, r is the number of rows, c is the number of columns, and n is the number of classes.
from keras import backend as K
def iou_coef(y_true, y_pred, smooth=1):
  intersection = K.sum(K.abs(y_true * y_pred), axis=[1,2,3])
  union = K.sum(y_true,[1,2,3])+K.sum(y_pred,[1,2,3])-intersection
  iou = K.mean((intersection + smooth) / (union + smooth), axis=0)
  return iou
  
iou_coef(im2, m3)
```
Erreur:
```
2020-06-05 00:48:26.050960: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-06-05 00:48:26.096634: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:1e:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2020-06-05 00:48:26.097152: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.097521: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.097727: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.097878: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.098051: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.098187: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.098325: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudnn.so.7'; dlerror: libcudnn.so.7: cannot open shared object file: No such file or directory
2020-06-05 00:48:26.098334: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-06-05 00:48:26.099037: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-06-05 00:48:26.110382: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 2400000000 Hz
2020-06-05 00:48:26.110945: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x45c9ce0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-06-05 00:48:26.110963: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-06-05 00:48:26.113305: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-06-05 00:48:26.113319: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in iou_coef
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 1900, in abs
    return tf.abs(x)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py", line 180, in wrapper
    return target(*args, **kwargs)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py", line 268, in abs
    return gen_math_ops._abs(x, name=name)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/tensorflow/python/ops/gen_math_ops.py", line 52, in _abs
    _ops.raise_from_not_ok_status(e, name)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 6653, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.NotFoundError: Could not find valid device for node.
Node:{{node Abs}}
All kernels registered for op Abs :
  device='XLA_GPU'; T in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_INT16, DT_INT8, DT_INT64, DT_BFLOAT16, DT_HALF]
  device='XLA_CPU'; T in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_INT16, DT_INT8, DT_INT64, DT_BFLOAT16, DT_HALF]
  device='XLA_CPU_JIT'; T in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_INT16, DT_INT8, DT_INT64, DT_BFLOAT16, DT_HALF]
  device='XLA_GPU_JIT'; T in [DT_FLOAT, DT_DOUBLE, DT_INT32, DT_INT16, DT_INT8, DT_INT64, DT_BFLOAT16, DT_HALF]
  device='GPU'; T in [DT_INT32]
  device='GPU'; T in [DT_INT64]
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_FLOAT]
  device='GPU'; T in [DT_HALF]
  device='CPU'; T in [DT_INT64]
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_INT16]
  device='CPU'; T in [DT_INT8]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_BFLOAT16]
  device='CPU'; T in [DT_HALF]
 [Op:Abs]
```

### Test 3rd implementation
Reference: <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_score.html>

Retourne une erreur
```
from sklearn.metrics import jaccard_score
jaccard_score(im2, m3, average='samples')
```
Erreur:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/sklearn/utils/validation.py", line 73, in inner_f
    return f(**kwargs)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 736, in jaccard_score
    pos_label)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 1250, in _check_set_wise_labels
    y_type, y_true, y_pred = _check_targets(y_true, y_pred)
  File "/localscratch/vincelf.8491172.0/pyenv36/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 98, in _check_targets
    raise ValueError("{0} is not supported".format(y_type))
ValueError: unknown is not supported
```
## Using open source
* <https://ilmonteux.github.io/2019/05/10/segmentation-metrics.html>
* <https://gist.github.com/ilmonteux/8340df952722f3a1030a7d937e701b5a>
* <https://stats.stackexchange.com/questions/273537/f1-dice-score-vs-iou>


## Using mathlab

* <https://docs.computecanada.ca/wiki/MATLAB>
* <https://www.mathworks.com/help/vision/semantic-segmentation.html?category=semantic-segmentation&s_tid=CRUX_gn_documentation_semantic-segmentation>

